src/GafferImage/OpenImageIOReader.cpp

```diff
@@ -0,0 +1,625 @@
+// files (expressions and timewarps etc can change the fileName at any time).
+// The OIIO image cache is designed for exactly this scenario.
+//////////////////////////////////////////////////////////////////////////
+
+namespace
+{
+
+spin_rw_mutex g_imageCacheMutex;
+ImageCache *imageCache()
+{
+	spin_rw_mutex::scoped_lock lock( g_imageCacheMutex, false );
+	static ImageCache *cache = 0;
+	if( cache == 0 )
+	{
+		if( lock.upgrade_to_writer() )
+		{
+			cache = ImageCache::create();
+			// ImageReaderTest.testOIIOJpgRead exposes a bug in
+			// OpenImageIO where ImageCache::get_pixels() returns
+			// incorrect data when reading from non-float images.
+			// By forcing the image to be float on loading, we
+			// can work around that problem.
+			/// \todo Consider removing this once the bug is fixed in
+			/// OIIO - and test any performance implications of performing
+			/// the conversion in get_pixels() rather than immediately on load.
+			cache->attribute( "forcefloat", 1 );
+
+			// Set an initial cache size of 500Mb
+			cache->attribute( "max_memory_MB", 500.0f );
+		}
+	}
+	return cache;
+}
+
+// Returns the OIIO ImageSpec for the given filename in the current
+// context. Throws if the file is invalid, and returns NULL if
+// the filename is empty.
+const ImageSpec *imageSpec( ustring &fileName, OpenImageIOReader::ErrorMode mode )
+{
+	if( fileName.empty() )
+	{
+		return NULL;
+	}
+	ImageCache *cache = imageCache();
+	const ImageSpec *spec = cache->imagespec( ustring( fileName.c_str() ) );
+	if( !spec )
+	{
+		if( mode == OpenImageIOReader::Black )
+		{
+			// we can simply return the null spec and rely on the
+			// compute methods to return default plug values.
+			return spec;
+		}
+		else if( mode == OpenImageIOReader::Hold )
+		{
+			const std::string &fileNameStr = fileName.string();
+			boost::filesystem::path dir = boost::filesystem::path( fileNameStr ).parent_path();
+			if ( dir.string() == "" )
+			{
+				dir = ".";
+			}
+			
+			std::vector<IECore::FileSequencePtr> sequences;
+			IECore::ls( dir.string(), sequences, /* minSequenceSize */ 1 );
+			for( std::vector<IECore::FileSequencePtr>::iterator it = sequences.begin(); it != sequences.end(); ++it )
+			{
+				IECore::FileSequence *sequence = it->get();
+				std::string prefix = ( dir / boost::filesystem::path( sequence->getPrefix() ) ).string();
+				sequence->setPrefix( prefix );
+				std::string suffix = sequence->getSuffix();
+				if (
+					( fileNameStr.substr( 0, prefix.size() ) == prefix ) &&
+					( fileNameStr.substr( fileNameStr.size() - suffix.size(), suffix.size() ) == suffix )
+				)
+				{
+					std::vector<IECore::FrameList::Frame> frames;
+					sequence->getFrameList()->asList( frames );
+					IECore::FrameList::Frame currentFrame = sequence->frameForFileName( fileNameStr );
+					std::vector<IECore::FrameList::Frame>::iterator fIt = std::lower_bound( frames.begin(), frames.end(), currentFrame );
```

Comment:
i ask that because we discuss it with lorenzo and in IEMedia Ehsan did a sustem to handle such a case.

If you dont have a valid first (expected) frame, we look for the first valid frame.

So say you expect a range starting at 1001 but the frame sequence your dealing with start at 1005, you hold 1005 during 1001 to 1004.

I ask you the question because Im not sure if this node has a concept of expected frame range, maybe it's more like, I deal with what I have and handle error that may rise in the frame range.



> On Nov 6, 2015, at 14:14, Andrew Kaufman <notifications@github.com> wrote:
> 
> In src/GafferImage/OpenImageIOReader.cpp:
> 
> > +			IECore::ls( dir.string(), sequences, /* minSequenceSize */ 1 );
> > +			for( std::vector<IECore::FileSequencePtr>::iterator it = sequences.begin(); it != sequences.end(); ++it )
> > +			{
> > +				IECore::FileSequence *sequence = it->get();
> > +				std::string prefix = ( dir / boost::filesystem::path( sequence->getPrefix() ) ).string();
> > +				sequence->setPrefix( prefix );
> > +				std::string suffix = sequence->getSuffix();
> > +				if (
> > +					( fileNameStr.substr( 0, prefix.size() ) == prefix ) &&
> > +					( fileNameStr.substr( fileNameStr.size() - suffix.size(), suffix.size() ) == suffix )
> > +				)
> > +				{
> > +					std::vector<IECore::FrameList::Frame> frames;
> > +					sequence->getFrameList()->asList( frames );
> > +					IECore::FrameList::Frame currentFrame = sequence->frameForFileName( fileNameStr );
> > +					std::vector<IECore::FrameList::Frame>::iterator fIt = std::lower_bound( frames.begin(), frames.end(), currentFrame );
> I guess you're wondering if it should fall back to the default format with an all black image instead of erroring?
> 
> —
> Reply to this email directly or view it on GitHub.
>